test/box/gh-5958-rolled-back-replace-background-built-index.test.lua

```diff
@@ -0,0 +1,268 @@
+    box.begin()
+    s:replace{0, 0, "valid"}
+    local svp = box.savepoint()
+    s:replace{1, 1, "changed D:"}
+    box.rollback_to_savepoint(svp)
+    box.commit()
+end;
+
+test_run:cmd("setopt delimiter ''");
+
+disturber = joinable(fiber.new(disturb))
+creator = joinable(fiber.new(create, true))
+
+disturber:join()
+creator:join()
+
+-- Check if replace was not rolled back.
+assert(s.index.pk:get{0} ~= nil and s.index.pk:get{0} == s.index.sk:get{0})
+-- Check if replace was rolled back.
+assert(s.index.pk:get{1} == origin[1] and s.index.pk:get{1} == s.index.sk:get{1})
+
+s.index.pk:delete{0}
+s.index.sk:drop()
+
+
+
+------ Suite to test rolled back changes in alter-in-background primary index  ------
+
+
+
+test_run:cmd("setopt delimiter ';'");
+
+-- Wait for fiber yield during building of index and then make all possible replaces:
+-- 1. replace nil -> x
+-- 2. replace x -> y
+-- 3. replace x -> nil
+-- and then roll them back.
+function disturb()
+    while not started do fiber.sleep(0) end
+    box.begin()
+    s:replace{0, 0, "invalid"}
+    s:replace{1, 1, "changed D:"}
+    s:delete{2}
+    box.rollback()
+end;
+
+-- Alter primary index in background.
+-- If primary index will not be altered in background, test will not be passed
+-- because it will run out of time (disturber:join() will wait forever).
+function alter()
+    started = true
+    i:alter({parts={{field = 2, type = 'unsigned'}}})
+    started = false
+end;
+
+test_run:cmd("setopt delimiter ''");
+
+disturber = joinable(fiber.new(disturb))
+alterer = joinable(fiber.new(alter))
+
+disturber:join()
+alterer:join()
+
+-- Check if changes were rolled back.
+assert(s.index.pk:get{0} == origin[0])
+assert(s.index.pk:get{1} == origin[1])
+assert(s.index.pk:get{2} == origin[2])
+
+
+
+------ Suite to check if replaced tuples are referenced by primary index after rollback  ------
+
+
+
+test_run:cmd("setopt delimiter ';'");
+
+-- Wait for fiber yield during altering of primary index and then replace some tuples.
+function disturb()
+    while not started do fiber.sleep(0) end
+    box.begin()
+    for i = 1, 100 do
+        s:replace{i, i, "new tuple must be referenced by primary key"}
+    end
+    box.rollback()
+end;
+
+-- Alter primary index in background.
+-- If primary index will not be altered in background, test will not be passed
+-- because it will run out of time (disturber:join() will wait forever).
+function alter()
+    started = true
+    i:alter({parts={{field = 1, type = 'unsigned'}}})
+    started = false
+end;
+
+test_run:cmd("setopt delimiter ''");
+
+disturber = joinable(fiber.new(disturb))
+alterer = joinable(fiber.new(alter))
+
+disturber:join()
+alterer:join()
+
+-- If some tuples are not referenced, garbage collector will delete them
+-- and we will catch segmentation fault on s:drop().
+collectgarbage("collect")
+s:drop()
+
+origin = nil
+started = nil
+
+
+
+------------------ Suite to make sure there is no memory leak  ------------------
```

Comment:
This looks like test from previous patch...

Reply:
In this test replaces are rolled back